File: macos/VideoRoom/Utilities/EditorRegistry.swift
import AppKit
import Foundation
import Combine

/// Per-user configuration for one editor. Persisted in `UserDefaults`.
struct EditorConfig: Equatable {
    /// Whether this editor shows up in the right-click "Open with" menu.
    var enabled: Bool
    /// Custom executable / `.app` bundle path (overrides catalog defaults).
    var customPath: String

    static let `default` = EditorConfig(enabled: true, customPath: "")
}

/// Detection + persistence + launch for external editors.
///
/// Detection is filesystem-only: checks `defaultPaths` plus any configured
/// custom path, returning the first one that exists. No sub-process calls.
///
/// Configuration is stored in `UserDefaults` under per-editor keys
/// (`editor.<id>.enabled`, `editor.<id>.path`). `ObservableObject` so SwiftUI
/// preferences views auto-refresh after each change.
@MainActor
final class EditorRegistry: ObservableObject {
    static let shared = EditorRegistry()

    /// Bumped after every mutation; views observing `@Published` properties
    /// will re-read the registry. Cheaper than emitting per-editor changes.
    @Published private(set) var revision: Int = 0

    private let defaults = UserDefaults.standard
    private let fm = FileManager.default

    /// Currently loaded config for one editor.
    func config(for editor: ExternalEditor) -> EditorConfig {
        // Default to enabled=true so newly-supported editors light up
        // automatically when the user has them installed.
        let enabledKey = "editor.\(editor.id).enabled"
        let pathKey = "editor.\(editor.id).path"
        let enabled = (defaults.object(forKey: enabledKey) as? Bool) ?? true
        let path = defaults.string(forKey: pathKey) ?? ""
        return EditorConfig(enabled: enabled, customPath: path)
    }

    /// Persist a config change for one editor.
    func update(_ editor: ExternalEditor, _ config: EditorConfig) {
        defaults.set(config.enabled, forKey: "editor.\(editor.id).enabled")
        if config.customPath.isEmpty {
            defaults.removeObject(forKey: "editor.\(editor.id).path")
        } else {
            defaults.set(config.customPath, forKey: "editor.\(editor.id).path")
        }
        revision += 1
    }

    // MARK: - Detection

    /// Resolve which path VideoRoom should launch for `editor`, or `nil` if it
    /// isn't installed anywhere we know to look.
    func resolvedPath(for editor: ExternalEditor) -> String? {
        let cfg = config(for: editor)
        if !cfg.customPath.isEmpty, fm.fileExists(atPath: cfg.customPath) {
            return cfg.customPath
        }
        let candidates = editor.defaultPaths[.macOS] ?? []
        return candidates.first(where: { fm.fileExists(atPath: $0) })
    }

    func isInstalled(_ editor: ExternalEditor) -> Bool {
        resolvedPath(for: editor) != nil
    }

    /// Editors that should appear in the right-click "Open with" menu right
    /// now — enabled by the user and present on disk.
    var availableEditors: [(editor: ExternalEditor, path: String)] {
        EditorCatalog.forCurrentPlatform.compactMap { editor in
            guard config(for: editor).enabled else { return nil }
            guard let path = resolvedPath(for: editor) else { return nil }
            return (editor, path)
        }
    }

    // MARK: - Launching

    /// Launch `editor` with the given file paths. On macOS this uses
    /// `NSWorkspace.open(_:withApplicationAt:configuration:completionHandler:)`
    /// which respects Launch Services, sandbox prompts, and Apple Events
    /// permission. Returns `true` on a best-effort spawn.
    @discardableResult
    func launch(_ editor: ExternalEditor, files: [String]) -> Bool {
        guard let path = resolvedPath(for: editor) else { return false }
        let appURL = URL(fileURLWithPath: path)
        let fileURLs: [URL] = editor.supportsFileArgs
            ? files.map { URL(fileURLWithPath: $0) }
            : []
        let config = NSWorkspace.OpenConfiguration()
        config.activates = true
        if fileURLs.isEmpty {
            NSWorkspace.shared.openApplication(at: appURL, configuration: config, completionHandler: nil)
        } else {
            NSWorkspace.shared.open(fileURLs, withApplicationAt: appURL, configuration: config, completionHandler: nil)
        }
        return true
    }

    /// Hand a file off to macOS's default video player.
    @discardableResult
    func openWithDefault(_ filePath: String) -> Bool {
        let url = URL(fileURLWithPath: filePath)
        guard fm.fileExists(atPath: filePath) else { return false }
        return NSWorkspace.shared.open(url)
    }

    /// Reveal a file in the Finder.
    @discardableResult
    func revealInFinder(_ filePath: String) -> Bool {
        let url = URL(fileURLWithPath: filePath)
        guard fm.fileExists(atPath: filePath) else { return false }
        NSWorkspace.shared.activateFileViewerSelecting([url])
        return true
    }

    /// Open the editor's homepage in the user's default browser.
    @discardableResult
    func openHomepage(_ editor: ExternalEditor) -> Bool {
        guard let url = URL(string: editor.homepage) else { return false }
        return NSWorkspace.shared.open(url)
    }

    // MARK: - Browse picker

    /// Show an NSOpenPanel for choosing an app bundle. Returns the picked
    /// path or `nil` if the user cancelled.
    func pickCustomPath(for editor: ExternalEditor) -> String? {
        let panel = NSOpenPanel()
        panel.title = "Choose \(editor.name)"
        panel.prompt = "Select"
        panel.canChooseFiles = true
        panel.canChooseDirectories = true   // .app bundles are technically dirs
        panel.allowsMultipleSelection = false
        panel.allowedContentTypes = []      // accept anything
        panel.directoryURL = URL(fileURLWithPath: "/Applications")
        if panel.runModal() == .OK, let url = panel.url {
            return url.path
        }
        return nil
    }
}

File: desktop/src/main/kotlin/com/videoroom/data/editors/EditorRegistry.kt
package com.videoroom.data.editors

import java.io.File
import java.util.Properties

/**
 * Per-user configuration for one editor — stored in [EditorRegistry]'s
 * properties file.
 */
data class EditorConfig(
    /** Whether this editor is enabled (shows up in the context menu). */
    val enabled: Boolean,
    /** Custom executable / .app path (overrides the catalog's [defaultPaths]). */
    val customPath: String,
) {
    companion object {
        val Default = EditorConfig(enabled = true, customPath = "")
    }
}

/**
 * Detection + persistence + launch for external editors.
 *
 * Detection is filesystem-only: it walks [ExternalEditor.defaultPaths] and any
 * configured custom path, returning the first one that exists. We deliberately
 * don't shell out to `which` or query system caches so detection is fast and
 * has no side effects.
 *
 * Configuration is persisted as a [java.util.Properties] file under
 * `~/.config/videoroom/external_editors.properties`. The format is plain
 * key/value, one editor per row:
 * ```
 * davinci-resolve.enabled=true
 * davinci-resolve.path=/Applications/DaVinci Resolve/DaVinci Resolve.app
 * ```
 */
class EditorRegistry {
    private val configFile: File by lazy {
        val home = System.getProperty("user.home") ?: "."
        val dir = File(home, ".config/videoroom")
        if (!dir.exists()) dir.mkdirs()
        File(dir, "external_editors.properties")
    }

    private val props = Properties().also { p ->
        if (configFile.exists()) {
            try {
                configFile.inputStream().use { p.load(it) }
            } catch (_: Exception) {
                // Ignore corrupt file; default config applies.
            }
        }
    }

    /** Load (or compute the default for) one editor's config. */
    fun configFor(editorId: String): EditorConfig {
        val enabled = props.getProperty("$editorId.enabled")
        val path = props.getProperty("$editorId.path") ?: ""
        // Default to enabled=true so newly-supported editors light up
        // automatically when they're installed.
        return EditorConfig(
            enabled = enabled?.toBooleanStrictOrNull() ?: true,
            customPath = path
        )
    }

    /** Persist one editor's config. Saves the file synchronously. */
    fun update(editorId: String, config: EditorConfig) {
        props.setProperty("$editorId.enabled", config.enabled.toString())
        if (config.customPath.isNotEmpty()) {
            props.setProperty("$editorId.path", config.customPath)
        } else {
            props.remove("$editorId.path")
        }
        save()
    }

    private fun save() {
        try {
            configFile.outputStream().use { props.store(it, "VideoRoom external editors") }
        } catch (_: Exception) {
            // Best-effort; persisted across sessions but not critical.
        }
    }

    // ────────────────────────────────────────────────────────────────────────
    // Detection
    // ────────────────────────────────────────────────────────────────────────

    /**
     * Resolve the path VideoRoom should launch for [editor], or null if the
     * editor isn't installed at any known location. The custom path (if set)
     * wins; otherwise we probe [ExternalEditor.defaultPaths] for the current
     * host.
     */
    fun resolvePath(editor: ExternalEditor): String? {
        val cfg = configFor(editor.id)
        if (cfg.customPath.isNotEmpty() && File(cfg.customPath).exists()) {
            return cfg.customPath
        }
        val host = HostPlatform.detect()
        val candidates = editor.defaultPaths[host].orEmpty()
        return candidates.firstOrNull { File(it).exists() }
    }

    /** True if [editor] is currently installed (or has a working custom path). */
    fun isInstalled(editor: ExternalEditor): Boolean = resolvePath(editor) != null

    /** Editors that should appear in the "Open with" context menu right now. */
    fun availableEditors(): List<Pair<ExternalEditor, String>> {
        return editorsForCurrentPlatform()
            .filter { configFor(it.id).enabled }
            .mapNotNull { editor ->
                resolvePath(editor)?.let { editor to it }
            }
    }

    // ────────────────────────────────────────────────────────────────────────
    // Launching
    // ────────────────────────────────────────────────────────────────────────

    /**
     * Launch [editor] with the given video file paths. Returns true if the
     * process spawned successfully; the caller can show an error toast on
     * false. If [ExternalEditor.supportsFileArgs] is false, the editor is
     * launched empty and the user must import manually.
     */
    fun launch(editor: ExternalEditor, files: List<String>): Boolean {
        val path = resolvePath(editor) ?: return false
        val argv = buildLaunchArgv(path, files, editor.supportsFileArgs)
        return try {
            ProcessBuilder(argv).inheritIO().start()
            true
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Build the platform-appropriate argv to launch [path] (an executable or
     * a macOS `.app` bundle) with [files].
     *
     *   * macOS .app bundle → `open -a "<path>" <files...>` so Launch Services
     *     handles the bundle correctly and any "is the developer trusted?"
     *     prompts show up.
     *   * Anything else → spawn the executable directly with the file paths.
     */
    private fun buildLaunchArgv(
        path: String,
        files: List<String>,
        supportsFileArgs: Boolean,
    ): List<String> {
        val isMacAppBundle = path.endsWith(".app") || path.contains(".app/")
        val filesToPass = if (supportsFileArgs) files else emptyList()
        return when {
            isMacAppBundle -> {
                val base = listOf("/usr/bin/open", "-a", path)
                if (filesToPass.isNotEmpty()) base + filesToPass else base
            }
            else -> listOf(path) + filesToPass
        }
    }

    /** Hand a file off to the OS's default video player. */
    fun openWithDefault(filePath: String): Boolean {
        return try {
            val f = File(filePath)
            if (!f.exists()) return false
            // Desktop.open uses the OS's default association for the file type.
            java.awt.Desktop.getDesktop().open(f)
            true
        } catch (_: Exception) {
            false
        }
    }

    /**
     * Reveal a file in the OS's native file manager (Finder / Explorer /
     * configured file manager on Linux).
     */
    fun revealInFileManager(filePath: String): Boolean {
        val f = File(filePath)
        if (!f.exists()) return false
        return try {
            when (HostPlatform.detect()) {
                HostPlatform.MacOS -> ProcessBuilder("/usr/bin/open", "-R", filePath).start()
                HostPlatform.Windows -> ProcessBuilder("explorer.exe", "/select,$filePath").start()
                HostPlatform.Linux -> {
                    // xdg-open the parent dir; no portable "select" equivalent.
                    val parent = f.parentFile?.absolutePath ?: return false
                    ProcessBuilder("xdg-open", parent).start()
                }
                HostPlatform.Unknown -> return false
            }
            true
        } catch (_: Exception) {
            false
        }
    }

    /** Open the editor's homepage in the user's default web browser. */
    fun openHomepage(editor: ExternalEditor): Boolean {
        return try {
            java.awt.Desktop.getDesktop().browse(java.net.URI(editor.homepage))
            true
        } catch (_: Exception) {
            false
        }
    }

    companion object {
        /** Shared single-process registry. */
        val Default by lazy { EditorRegistry() }
    }
}
